celery/utils/functional.py

```diff
@@ -266,7 +266,11 @@ def head_from_fun(fun, bound=False, debug=False):
     # in pure-Python.  Instead we use exec to create a new function
     # with an empty body, meaning it has the same performance as
     # as just calling a function.
-    if not inspect.isfunction(fun) and hasattr(fun, '__call__'):
+    is_function = inspect.isfunction(fun)
+    is_callable = hasattr(fun, '__call__')
```

Comment:
I did not want to interfere with the logic here, as generally I believe it is not easy to detect functions correctly.

For example, one difference I can think is with old-style classes:

```
class A:
    pass

print callable(A) # True
print hasattr(A, '__call__') # False - would be True for new-style classes
```